File: src/backend/langgraph/adapters/providerAdapter.ts
import type { LoopContext } from "src/loopContext";
import type { Provider, ProviderEvent } from "src/services/api/types";
import type { Message } from "src/types/message";
import { getAgentTools } from "src/tools";
import { buildSystemPrompt } from "src/context";

export interface ProviderTurn {
  text: string;
  toolUses: { id: string; name: string; argsPartial: string }[];
  stopReason: "end_turn" | "tool_use" | "max_tokens";
  usage?: { inputTokens: number; outputTokens: number };
}

export interface StreamedEvent {
  type: "assistant_text" | "usage" | "error";
  text?: string;
  inputTokens?: number;
  outputTokens?: number;
  error?: Error;
}

/**
 * Drive the existing Provider asynchronously, emitting streamed LoopEvent-equivalents
 * through `onEvent`, and returning the structured turn for the graph node.
 */
export async function driveProvider(params: {
  provider: Provider;
  messages: Message[];
  ctx: LoopContext;
  model?: string;
  onEvent: (e: StreamedEvent) => void;
  signal: AbortSignal;
}): Promise<ProviderTurn> {
  const tools = getAgentTools();
  const system = buildSystemPrompt(params.ctx.stateStore.summary());
  const events = params.provider.stream({
    messages: params.messages,
    tools: tools.map((t) => ({
      name: t.name,
      description: t.description,
      inputSchema: t.inputSchema,
    })),
    system,
    ...(params.model ? { model: params.model } : {}),
    signal: params.signal,
  });

  let text = "";
  const toolUses: { id: string; name: string; argsPartial: string }[] = [];
  let stopReason: ProviderTurn["stopReason"] = "end_turn";
  let usage: { inputTokens: number; outputTokens: number } | undefined;

  for await (const ev of events) {
    if (params.signal.aborted) {
      const turn: ProviderTurn = { text, toolUses, stopReason: "end_turn" };
      if (usage !== undefined) turn.usage = usage;
      return turn;
    }
    switch ((ev as ProviderEvent).type) {
      case "text_delta":
        text += (ev as { text: string }).text;
        params.onEvent({ type: "assistant_text", text: (ev as { text: string }).text });
        break;
      case "tool_use_start":
        toolUses.push({ id: (ev as { id: string; name: string }).id, name: (ev as { id: string; name: string }).name, argsPartial: "" });
        break;
      case "tool_use_delta": {
        const d = ev as { id: string; argsPartialJson: string };
        const u = toolUses.find((t) => t.id === d.id);
        if (u) u.argsPartial += d.argsPartialJson;
        break;
      }
      case "tool_use_stop":
        break;
      case "message_stop":
        stopReason = (ev as { stopReason: ProviderTurn["stopReason"] }).stopReason;
        break;
      case "usage":
        usage = { inputTokens: (ev as { inputTokens: number }).inputTokens, outputTokens: (ev as { outputTokens: number }).outputTokens };
        params.onEvent({ type: "usage", ...usage });
        break;
      case "error": {
        const err = (ev as { error: Error }).error;
        params.onEvent({ type: "error", error: err });
        const turn: ProviderTurn = { text, toolUses, stopReason: "end_turn" };
        if (usage !== undefined) turn.usage = usage;
        return turn;
      }
    }
  }
  const turn: ProviderTurn = { text, toolUses, stopReason };
  if (usage !== undefined) turn.usage = usage;
  return turn;
}

File: src/docker_agent/engine/adapters/provider_adapter.py
"""Drive a Provider stream and collect a structured turn."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from docker_agent.core.prompt_builder import build_system_prompt
from docker_agent.services.api.types import (
    CallModelParams,
    ErrorEvent,
    MessageStopEvent,
    Provider,
    TextDeltaEvent,
    ToolSchema,
    ToolUseDeltaEvent,
    ToolUseStartEvent,
    ToolUseStopEvent,
    UsageEvent,
)
from docker_agent.types.message import Message


@dataclass
class ProviderTurn:
    text: str = ""
    tool_uses: list[dict[str, str]] = field(default_factory=list)
    stop_reason: str = "end_turn"
    usage: dict[str, int] | None = None


@dataclass
class StreamedEvent:
    type: str
    text: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    error: Exception | None = None


async def drive_provider(
    *,
    provider: Provider,
    messages: list[Message],
    ctx: Any,
    model: str | None = None,
    on_event: Callable[[StreamedEvent], None],
    signal: Any | None = None,
    tools: list[Any] | None = None,
) -> ProviderTurn:
    del ctx
    available_tools = tools or []
    params = CallModelParams(
        messages=messages,
        tools=[
            ToolSchema(name=t.name, description=t.description, input_schema=t.input_schema)
            for t in available_tools
        ],
        system=build_system_prompt(""),
        model=model,
    )

    turn = ProviderTurn()
    async for ev in provider.stream(params):
        if signal is not None and getattr(signal, "is_set", lambda: False)():
            return turn
        if isinstance(ev, TextDeltaEvent):
            turn.text += ev.text
            on_event(StreamedEvent(type="assistant_text", text=ev.text))
        elif isinstance(ev, ToolUseStartEvent):
            turn.tool_uses.append({"id": ev.id, "name": ev.name, "args_partial": ""})
        elif isinstance(ev, ToolUseDeltaEvent):
            for use in turn.tool_uses:
                if use["id"] == ev.id:
                    use["args_partial"] += ev.args_partial_json
        elif isinstance(ev, ToolUseStopEvent):
            pass
        elif isinstance(ev, MessageStopEvent):
            turn.stop_reason = ev.stop_reason
        elif isinstance(ev, UsageEvent):
            turn.usage = {"input_tokens": ev.input_tokens, "output_tokens": ev.output_tokens}
            on_event(
                StreamedEvent(
                    type="usage",
                    input_tokens=ev.input_tokens,
                    output_tokens=ev.output_tokens,
                )
            )
        elif isinstance(ev, ErrorEvent):
            on_event(StreamedEvent(type="error", error=ev.error))
            return turn
    return turn


__all__ = ["ProviderTurn", "StreamedEvent", "drive_provider"]
